devito/dse/aliases.py

```diff
@@ -36,244 +38,246 @@ def collect(exprs):
         a[i+2] - b[i+2] : because at least one operation differs
         a[i+2] + b[i] : because distance along ``i`` differ (+2 and +0)
     """
-    ExprData = namedtuple('ExprData', 'aindices offsets')
-
-    # Discard expressions:
-    # - that surely won't alias to anything
-    # - that are non-scalar
-    candidates = OrderedDict()
+    # Determine the potential aliases
+    candidates = []
     for expr in exprs:
-        if expr.lhs.is_Indexed:
-            continue
-        indexeds = retrieve_indexed(expr.rhs, mode='all')
-        if indexeds and not any(q_indirect(i) for i in indexeds):
-            handle = calculate_offsets(indexeds)
-            if handle:
-                candidates[expr.rhs] = ExprData(*handle)
-
-    aliases = OrderedDict()
-    mapper = OrderedDict()
+        candidate = analyze(expr)
+        if candidate is not None:
+            candidates.append(candidate)
+
+    # Group together the aliasing expressions (ultimately build an Alias for each
+    # group of aliasing expressions)
+    aliases = Aliases()
     unseen = list(candidates)
     while unseen:
+        c = unseen.pop(0)
+
         # Find aliasing expressions
-        handle = unseen.pop(0)
-        group = [handle]
-        for e in list(unseen):
-            if compare(handle, e):
-                if is_translated(candidates[handle].offsets, candidates[e].offsets):
-                    group.append(e)
-                    unseen.remove(e)
-
-        # Try creating a basis for the aliasing expressions' offsets
+        group = [c]
+        for i in list(unseen):
+            if compare_ops(c.expr, i.expr) and is_translated(c, i):
+                group.append(i)
+                unseen.remove(i)
+
+        # Try creating a basis spanning the aliasing expressions' iteration vectors
         try:
-            offsets = [tuple(candidates[e].offsets) for e in group]
-            COM, distances = calculate_COM(offsets)
-        except DSEException:
-            # Ignore these potential aliases and move on
+            COM, distances = calculate_COM(group)
+        except ValueError:
+            # Ignore these aliasing expressions and move on
             continue
 
-        alias = create_alias(handle, COM)
-
-        # An alias has been created, so I can now update the expression mapper
-        mapper.update([(i, group) for i in group])
+        # Create an alias expression centering `c`'s Indexeds at the COM
+        subs = {i: i.function[[x + c.fromlabel(x, 0) for x in b]]
+                for i, b, c in zip(c.indexeds, c.bases, COM)}
```

Comment:
this is super hard to understand. 
is it possible to separate it over more lines?